spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/core/BigQueryTemplate.java

```diff
@@ -0,0 +1,180 @@
+	 * @param taskScheduler the {@link TaskScheduler} used to poll for the status of
+	 *     long-running BigQuery operations
+	 */
+	public BigQueryTemplate(BigQuery bigQuery, String datasetName, TaskScheduler taskScheduler) {
+		Assert.notNull(bigQuery, "BigQuery client object must not be null.");
+		Assert.notNull(datasetName, "Dataset name must not be null");
+		Assert.notNull(taskScheduler, "TaskScheduler must not be null");
+
+		this.bigQuery = bigQuery;
+		this.datasetName = datasetName;
+		this.taskScheduler = taskScheduler;
+	}
+
+	/**
+	 * Sets whether BigQuery should attempt to autodetect the schema of the data when loading
+	 * data into an empty table for the first time. If set to false, the schema must be
+	 * defined explicitly for the table before load.
+	 * @param autoDetectSchema whether data schema should be autodetected from the structure
+	 *     of the data. Default is true.
+	 */
+	public void setAutoDetectSchema(boolean autoDetectSchema) {
+		this.autoDetectSchema = autoDetectSchema;
+	}
+
+	/**
+	 * Sets the {@link WriteDisposition} which specifies how data should be inserted into
+	 * BigQuery tables.
+	 * @param writeDisposition whether to append to or truncate (overwrite) data in the
+	 *     BigQuery table. Default is {@code WriteDisposition.WRITE_APPEND} to append data to
+	 *     a table.
+	 */
+	public void setWriteDisposition(WriteDisposition writeDisposition) {
+		Assert.notNull(writeDisposition, "BigQuery write disposition must not be null.");
+		this.writeDisposition = writeDisposition;
+	}
+
+	/**
+	 * Sets the {@link Duration} amount of time to wait between successive polls on the status
+	 * of a BigQuery job.
+	 * @param jobPollInterval the {@link Duration} poll interval for BigQuery job status
+	 *     polling
+	 */
+	public void setJobPollInterval(Duration jobPollInterval) {
+		Assert.notNull(jobPollInterval, "BigQuery job polling interval must not be null");
+		this.jobPollInterval = jobPollInterval;
+	}
+
+	@Override
+	public ListenableFuture<Job> writeDataToTable(
+			String tableName, InputStream inputStream, FormatOptions dataFormatOptions) {
+		TableId tableId = TableId.of(datasetName, tableName);
+
+		WriteChannelConfiguration writeChannelConfiguration = WriteChannelConfiguration
+				.newBuilder(tableId)
+				.setFormatOptions(dataFormatOptions)
+				.setAutodetect(this.autoDetectSchema)
+				.setWriteDisposition(this.writeDisposition)
+				.build();
+
+		TableDataWriteChannel writer = bigQuery.writer(writeChannelConfiguration);
+
+		try (OutputStream sink = Channels.newOutputStream(writer)) {
+			// Write data from data input file to BigQuery
+			StreamUtils.copy(inputStream, sink);
+		}
+		catch (IOException e) {
+			throw new BigQueryException("Failed to write data to BigQuery tables.", e);
+		}
+
+		if (writer.getJob() == null) {
+			throw new BigQueryException(
+					"Failed to initialize the BigQuery write job in message handler: " + this);
```

Comment:
`this`? It will always be the template class.
```suggestion
					"Failed to initialize the BigQuery write job.");
```

Reply:
oops, nice catch.